lib/gollum-lib/file.rb

```diff
@@ -41,24 +44,32 @@ def path_compare(query, match_path, hyphened_tags, case_insensitive)
     # version - The String version ID to find.
     # try_on_disk - If true, try to return just a reference to a file
     #               that exists on the disk.
-    # global_match - If true, find a File matching path's filename, but not it's directory (so anywhere in the repo)
+    # global_match - If true, find a File matching path's filename, but not its directory (so anywhere in the repo)
     #
     # Returns a Gollum::File or nil if the file could not be found. Note
     # that if you specify try_on_disk=true, you may or may not get a file
     # for which on_disk? is actually true.
     def self.find(wiki, path, version, try_on_disk = false, global_match = false)
-      map = wiki.tree_map_for(version.to_s)
-
-      query_path = Pathname.new(::File.join(['/', wiki.page_file_dir, path].compact)).cleanpath.to_s
-      query_path.sub!(/^\/\//, '/') if Gem.win_platform? # On Windows, Pathname#cleanpath will leave double slashes at the start of a path intact, so sub them out.
+      query_path =  self.canonical_path(path, page_file_dir: wiki.page_file_dir)
+      path_segments = query_path.split('/')
+      filename = path_segments.last
+      dir = path_segments[0..-2].join('/')
 
-      begin
-        entry = map.detect do |entry|
-          path_match(query_path, entry, global_match, wiki.hyphened_tag_lookup, wiki.case_insensitive_tag_lookup)
+      if global_match && self.respond_to?(:global_find) # Only implemented for Gollum::Page
+        return self.global_find(wiki, version, query_path, try_on_disk) 
+      else
+        begin
```

Comment:
Similarly, I am thinking it may make sense to refactor this code so that it mostly becomes a method of `GitAccess` (so that we would hear call e.g. `wiki.access.find_blob_in_tree(dir, filename) {|blob_name| path_compare(filename, blob_name, ...}`.

Reply:
This turns out to be more complicated than I thought, as there is currently no `attr_accessor` for `wiki.access`. So either we have to make `wiki.access` public, or the `find_blob` method should be implemented as an instance method on `Wiki` that itself calls the underlying `GitAccess` instance, as e.g. `Wiki#tree_list` already does. Any opinion about what would make for the best API?

Reply:
As additional info, the call in `file.rb` would look something like this:

```
        entry = wiki.access.find_blob_in_tree(wiki.commit_for(version.to_s), dir.to_s) do |blob_name|
```

Where `wiki.commit_for` itself invokes the `GitAccess` class. This feels a bit ugly.

Reply:
After discussion we've concluded this change is too complicated and not worth it, so we're leaving the find file logic in`file.rb` as opposed to `git_access.rb`